keiko/keiko/keiko.py

```diff
@@ -135,32 +164,41 @@ def generate_report(
 
         # run pdflatex
         cmd = [
-            "pdflatex",
+            "latexmk",
+            "-xelatex",
             "-synctex=1",
             "-interaction=nonstopmode",
             preprocessed_tex_path.as_posix(),
         ]
         env = {**os.environ, "TEXMFVAR": directory}
-        for i in (1, 2):
-            output = subprocess.run(cmd, cwd=directory, env=env, capture_output=True, check=False)
-            current_span.add_event(f"Completed pdflatex run {i}")
+
+        def log_output(level, output):
+            if not logger.isEnabledFor(level):
+                return
```

Comment:
I want to avoid calling the log function for each line of output when not needed, so I think this is the cleanest way to do so?

Reply:
Yes it makes sense to not run the whole for loop when it is not necessary.

The function doesn't use variables from the outer function so I can see no reason why it should be defined inside the function. Moving it to the module scope means the function will only be defined once on import time instead of everytime the function is called.

Reply:
> The function doesn't use variables from the outer function so I can see no reason why it should be defined inside the function. Moving it to the module scope means the function will only be defined once on import time instead of everytime the function is called.

I'm using the `report_id` now, but could still move it out of the function if that is preferred.

Reply:
I don't think the extra if statement makes the code more complex. And while the overhead in this single case might be insignificant, all those little overheads everywhere do add up. I also think it is a good thing to optimize our code so we also do it when it does matter even when we might nog expect it to.

With regards to the nested function, it is also a very small optimization to not nest them and I also think nesting often makes it a bit harder to understand what is going, but I don't feel very strong about it. I'm moving the ticket to QA.